pycparser/_ast_gen.py

```diff
@@ -253,21 +283,35 @@ def visit_Constant(self, node):
         *   Modeled after Python's own AST visiting facilities
             (the ast module of Python 3.0)
     """
+    
+    _method_cache = None
+    
     def visit(self, node):
         """ Visit a node.
         """
-        method = 'visit_' + node.__class__.__name__
-        visitor = getattr(self, method, self.generic_visit)
+        
+        method_cache = self._method_cache
+        if method_cache is None:
+            method_cache = {}
+            self._method_cache = method_cache
+            
+        visitor = method_cache.get(node.__class__.__name__, None)
```

Comment:
Can you use dict's `setdefault` here?

Reply:
That would be functionally equivalent, but performance would be similar to the original implementation, as we would still compute/retrieve the visitor for each call.

Using `collections.defaultdict` is closer to the intended improvement: It creates a value only when the key is missing. However, it requires deriving a class from `defaultdict` and implementing `__missing__`.
As we need to access the current class' attributes, we need to either store a reference to it in the class, or use a closure.

I implemented both, and was surprised to see no improvement:
```
10000 iterations: 15.272s on Original
10000 iterations: 11.264s on Method cache only, using class name as key
10000 iterations: 17.910s on Method cache only, using setdefault
10000 iterations: 11.855s on Method cache only, using defaultdict
```

So `setdefault` is not better, and neither is `defaultdict`.

Reply:
OK, thanks